Replace bind with a version that uses const references

bremoran opened the pull request:
This version of bind does away with stdarg and uses a friend class instead.  This allows FunctionPointer to work with all types of argument, including references.

bremoran commented:
cc @bogdanm @niklas-arm @pan- @autopulated

autopulated commented:
:thumbsup: looks much nicer, can't see anything obviously wrong – notice we still don't have tests checking the lifetime of argument objects though?

bremoran commented:
Yes, we definitely need more extensive testing.  I have written some more tests, but they don't go to that depth yet either.

bogdanm commented:
I'll try to run the Event test on this change (and maybe automate the test a bit more in the process) since it touches some lifetime aspects, but if there's not enough confidence, it's safer to merge this after the TechCon release.

bremoran commented:
Since github hides commit messages by default, here's the message from the previous commit:

* The previous code has a bug, which prevented minar from working.  In bind, the FunctionPointer’s member caller, object, and member were never copied to the new FunctionPointerBind.  This meant that no bound pointers would work.  This is corrected by adding a copy constructor for initializing a FunctionPointerBind with a FunctionPointerBase.

* Added numerous tests to validate the new implementation.  There are still too many copies of arguments.  A single FunctionPointer call results in 3 new instances of each argument.  This could be reduced with some additional work.

* Removed the indirection from ```operator()``` to ```call()``` since this introduces one additional copy of each argument.

* Added documentation about yotta config options.

autopulated commented on test/FunctionPointer/lifetime.cpp:
```diff
@@ -0,0 +1,138 @@
+    LifetimeChecker(int arg) : _arg(arg) {
+        _activeInstances++;
+    }
+    LifetimeChecker(const LifetimeChecker & lc) : _arg(lc._arg) {
```
should probably also define a copy-assignment operator, in case a future implementation uses that (https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)#Rule_of_Three)

autopulated commented on test/FunctionPointer/main.cpp:
```diff
@@ -54,20 +59,22 @@ void runTest(void) {
+        result = result && ecp_flag;
         printf("ecp_flag = %d\r\n", ecp_flag);
     }
+    result = result && testInts();
```
this pattern is going to skip the evaluation of subsequent tests if result is already false – do you want that?

bremoran replied:
I was aware of that, but it's less than ideal.  I'll do something slightly better.

bremoran commented:
Bump for review

bogdanm commented on core-util/FunctionPointer.h:
```diff
@@ -83,10 +83,14 @@ class FunctionPointer0 : public FunctionPointerBase<R>{
-        FunctionPointerBind<R> fp;
-        fp.bind(&FunctionPointerBase<R>::_nullops, (ArgStruct *) NULL, this);
+        FunctionPointerBind<R> fp(*this);
+        void * storage = this->preBind(fp, (ArgStruct *)NULL, &FunctionPointerBase<R>::_nullops);
```
Please don't use `camelCase`

bremoran replied:
Why not?  minar is camelCase.

bogdanm replied:
We've been though this. Yes, MINAR is the pretty much the only part which is camelCase, the vast majority of mbed OS is not. Let's stick to our regular coding style, we'll fix MINAR later.

bremoran replied:
fluent APIs with c-style snake-case are going to look awfully weird...

autopulated replied:
yeah my recollection was that we had agreed to use camelCase for C++ APIs? @hugovincent

bremoran replied:
Also, preBind is an internal-only API.

bogdanm replied:
@autopulated, I am not aware of any such agreement.
@bremoran:
- fluent APIs are going to look absolutely OK IMO
- internal only doesn't mean that we can use a different coding style

bogdanm commented:
Looks good. Let's merge it and check if anything breaks.
